### PR TITLE
[Console] enable describing commands in ways that make the `list` command lazy

### LINKED ISSUE #33804: [Console] Distinguish definition of a command from its execution
**Description**  
Currently a command is a single PHP class, which implements both: its definition and the code to execute it. This is perfectly fine for most use cases, but has downsides for more complicated ones.

For commands that implement business logic and makes heavy use of dependency injection, creating the command object can be very expensive.

This fact was greatly mitigated by introducing `\Symfony\Component\Console\CommandLoader\CommandLoaderInterface`. Using that, will lazily instantiate only the command object specified, not all command objects like before. 
However we still face this performance impact when listing all commands (`console list`), where again all command objects are built "just" to show their names and the description (access the definition).

Therefore I'd propose to *optionally* separate definition from execution, by allowing commands to implement an interface which could roughly look like (don't pick on the names or details it's just an outline):

```php
<?php
namespace Symfony\Component\Console\Command;

interface CommandDefinitionInterface
{
    /**
     * @return CommandDefinition Includes InputDefinition, name, description, help...
     */
    public function getDefinition(): CommandDefinition;

    /**
     * Could also be part of CommandDefinition instead
     * 
     * @return string Class name or service name, which implements CommandExecutorInterface or is closure
     */
    public function getExecutor(): string;
}

interface CommandExecutorInterface
{
    /**
     * Executes command logic.
     *
     * @return int|null null or 0 if everything went fine, or an error code
     */
    public function execute(InputInterface $input, OutputInterface $output);
}
```

The benefit of this (compared to `setCode`, which could be deprecated then) is, that `ContainerCommandLoader` could be made responsible for creating the executor instead of the command object having to have a dependency to a DI container.

Any thoughts? Thanks!

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #33804
| License       | MIT
| Doc PR        | -

This PR improves the way one can describe a command so that the `list` command can be made lazy:
- when provided using the `$defaultName` property or the `console.command` tag, the name of a command is now exploded using the `|` character. The first name in the list defines the name of the command, the other ones its aliases. When the first name is the empty string, the second name is used instead, and the command is declared as hidden.
- a new `$defaultDescription` static property and a new `description` tag attribute allow for defining the commands' description while registering them.

Together, this is enough to make the `list` command lazy, because this command only accesses each command's name, aliases, hidden-status, and description.

On the implementation side, this PR adds a `LazyCommand` class that proxies regular commands to make them lazy for the target purpose.

This PR will enable support for attributes for configuring a command name+description+etc.
e.g. using the concepts in #39804:
`#[CommandAutoTag(name: 'foo:bar', desc: 'boo', hidden: true)]#`

The attribute could very well split the `hidden` and `aliases` settings apart - while the underlying code and pre-PHP8 apps would use the compact form, because dealing with many static properties + methods would be a maintenance pain imho.

### CODE REVIEW

**Comment by chalasr** [comment]
Looks really nice, thank you.
I'm going to challenge the new convention for hidden commands as I had to stare at it for a while before getting what it means. I'm not sure that the command visibility should be defined through its name, at least it's not obvious that an empty string before the first separator should be the way to control it. 
Could we use a special char in front of the name (`.` ?) or add another static prop+tag attr?

**Comment by nicolas-grekas** [comment]
Not hidden:
`$defaultName = 'app:foo|foo'`

Hidden:
`$defaultName = '|app:foo|goo'`

I don't think this is worth adding another static property.

**Review by chalasr — APPROVED** [review]

**Comment by javiereguiluz** [comment]
I don't like much the proposed convention to define a command as hidden because it's not self-explanatory.

As an alternative, would it be possible to store the list of commands in the cache dir while "compiling" the app? Same as we do with routes. Thanks.

**Comment by nicolas-grekas** [comment]
IMHO, there is no need to optimize for setting the "hidden" flag, as that's not a mainstream feature. It's readable enough as is.

> would it be possible to store the list of commands in the cache dir while "compiling" the app? Same as we do with routes. Thanks.

Sorry, I don't get what you mean, can you please describe your idea a bit more? What's the DX that goes with this?

**Comment by noniagriconomie** [comment]
nice to have this lazy feature to be added :)

my 2cts, why not `.app:foo|foo` for hidden? as a hidden Unix file for example

that way:

- the `|` for separating command name first and aliases next
- the `.` to hide the command

**Comment by jderusse** [comment]
Instead of adding static property for everything, what's about making the method `configure` static?.

```php
class AppCommand extends Command implements StaticCommandDescribedInterface
{
  public static function configure(): CommandDescription
  {
    return new CommandDescription('app:test')
      ->setDefinition()
      ->setHelp()
      ->addInput(...)
    ;
  }
}
```

class that can't be static and need internal service to resolve could implement another interface (that is incompatible with the previous)
```php
class AppCommand extends Command implements InstanceCommandDescribedInterface
{
  public function configure(): CommandDescription
  {
    return new CommandDescription('app:test')
      ->setDefinition()
      ->setHelp($this->getHint())
      ->addInput(...)
    ;
  }
}
```

**Comment by nicolas-grekas** [comment]
@jderusse that would kill the possibility to make the command configuration sensitive to constructor argument. Aka that could kill a lot of flexibility in the class. Yes, we don't use/need that in typical Symfony DI, but the console component is very much used standalone. We should be sure to not degrade it's DX when used outside of a typical Symfony stack.

From a SOLID pov, static methods are just bad.

**Comment by nicolas-grekas** [comment]
> why not `.app:foo|foo` for hidden

because it's not better than the current proposal?

**Comment by jderusse** [comment]
> @jderusse that would kill the possibility to make the command configuration sensitive to constructor argument. Aka that could kill a lot of flexibility in the class. Yes, we don't use/need that in typical Symfony DI, but the console component is very much used standalone. We should be sure to not degrade it's DX when used outside of a typical Symfony stack.
> 
> From a SOLID pov, static methods are just bad.

Isn't it addressed by the second part of my comment?

**Comment by nicolas-grekas** [comment]
@jderusse to not break BC, we should rather make static version opt-in. I don't know where this could lead, in terms of BC impact for example. Of course, feel free to give it a try, as I did here with my proposal.

**Comment by natewiebe13** [comment]
> > why not `.app:foo|foo` for hidden
> 
> because it's not better than the current proposal?

I think it's more clear as to what the intention is. Prefixing with a dot is common convention for hiding a file/directory in Linux and Mac OS. Whereas using the separator as a prefix, it isn't clear without first reading documentation that it's hiding the command instead of just a typo.

**Comment by stof** [comment]
@natewiebe13 but this reserves 2 chars (`.` and `|`) which cannot be used anymore in command names. `|` is fine as a reserved char is fine, as no one would use that char in their command name: `|` is a special char in all shells, and so using the command would require escaping it all the time.
But `.` has no special meaning and could totally be used in command names (even though that's not the recommended Symfony naming convention).

**Comment by stof** [comment]
> Isn't it addressed by the second part of my comment?

but that forbids lazy-loading these commands then, while the use case for relying in the constructor arguments in the configuration is much more likely to be related to configuring options or arguments than the description.

**Comment by natewiebe13** [comment]
> @natewiebe13 but this reserves 2 chars (`.` and `|`) which cannot be used anymore in command names. `|` is fine as a reserved char is fine, as no one would use that char in their command name: `|` is a special char in all shells, and so using the command would require escaping it all the time.
> But `.` has no special meaning and could totally be used in command names (even though that's not the recommended Symfony naming convention).

But wouldn't that just be for the first char in a command name? I wouldn't expect any periods in the command name to cause issues, just if the name was *prefixed* by a period.

**Edit**
For example, I'd still think `.app.do-something` would still be valid as `app.do-something`. It would just be hidden.

**Comment by stof** [comment]
@nicolas-grekas there is still the issue of `Command::isEnabled()` though. It impacts `list` as well.

**Comment by nicolas-grekas** [comment]
> there is still the issue of Command::isEnabled() though. It impacts list as well.

Isn't this covered? See the implementation of `isEnabled()`. Otherwise, I may have missed what you meant.

**Comment by nicolas-grekas** [comment]
> I wouldn't expect any periods in the command name to cause issues, just if the name was prefixed by a period.

Until one will report a BC break :) This happens all the time when we do such assumptions.

Honestly, the leading pipe is fine, it's as readable. Everything needs to be taught anyway.

**Comment by stof** [comment]
> Isn't this covered? See the implementation of `isEnabled()`. Otherwise, I may have missed what you meant.

disabled commands are excluded by `list`

**Comment by ro0NL** [comment]
@stof pipe char may be quoted 😅 im fine with it, otherwise `static $defaultAliases & co`

**Comment by nicolas-grekas** [comment]
> disabled commands are excluded by list

yep, and this is managed by a call to `$command->isEnabled()` when adding them to the app. I think this works as expected since `LazyCommand` will return `true` without instantiating the real command.

**Comment by stof** [comment]
@nicolas-grekas that's precisely the issue. If LazyCommand always returns true, how does the feature actually work for commands overriding the method to use the feature ?

**Comment by nicolas-grekas** [comment]
> If LazyCommand always returns true, how does the feature actually work for commands overriding the method to use the feature ?

Oh, OK. Easy then :) If one wants to disable a command in a Symfony app, on should use DI (eg disabling autoconfiguration, removing the tag). I think `isEnabled()` is useful for situations where DI is not in use. Alternatively, one would *not* set the description using the new way, and this would fall back to regular non-lazy registration.

**Comment by ro0NL** [comment]
@nicolas-grekas should we make the effort for full compat between DI and regular configuration? Meaning adding a bunch of DI tag attributes for description etc.

#39804 then enables declarative syntax: `#[CommandAutoTag(name: 'foo:bar', desc: 'boo', hidden: true)]#` 

i tend to fancy it a bit more than our current "static prop" solution.

edit: it does become a complex thing given we could leverage CommandAutoTag without autoconfig being applied in DI. But generally spoken attribute seems to scale well for this.

File: src/Symfony/Component/Console/Command/Command.php
Comment (rheh) [inline src/Symfony/Component/Console/Command/Command.php:586]: Question: Do we need the is_array?  We are now building $list and it is going to be an array.  I think the check is ensuring $this->aliases is set to an array.  As $list is always an array and it is guaranteed to contain the same as $aliases can we just use that all the time?
```
 $this->aliases = $list;
```

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (rheh) [inline src/Symfony/Component/Console/Command/LazyCommand.php]: I believe all these setters return $this which implies the intent to chain the calls?

```
  $this->setName($name)
     ->setAliases($aliases)
     ->setHidden($isHidden)
     ->setDescription($description);
```

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (rheh) [inline src/Symfony/Component/Console/Command/LazyCommand.php:190]: Can this variable be anything but \Closure it is mandatory and not nullable on the constructor?

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php
Comment (rheh) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php:114]: Question: Is variable interpolation preferred over sprintf?

```
  throw new InvalidArgumentException("Class \"{$class}\" used for service \"{$id}\" cannot be found.');
```

Comment (stof) [reply src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php:114]: symfony prefers using `sprintf` for such case

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/LazyCommand.php]: updated

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/LazyCommand.php:190]: yes, see the line below this one, where the variable is resolved into a `Command`

File: src/Symfony/Component/Console/Command/Command.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/Command.php:586]: This is an optimization trick that saves memory: when the variable is an array, $list is a hard copy of $aliases, while $aliases is a copy-on-write reference.

**Comment by nicolas-grekas** [comment]
> should we make the effort for full compat between DI and regular configuration? Meaning adding a bunch of DI tag attributes for description etc.
> #39804 then enables declarative syntax: #[CommandAutoTag(name: 'foo:bar', desc: 'boo', hidden: true)]#

Yes, after this one and #39804 are merged, we'll be able to provide support for attributes for configuring commands, with or without DI.

> tend to fancy it a bit more than our current "static prop" solution.

the solution will rely on static methods that will read the attributes using reflection (the properties are a local implementation detail).

**Comment by nicolas-grekas** [comment]
To make this explicit to everyone reading here:

This PR will enable support for attributes for configuring a command name+description+etc.
e.g. using the concepts in #39804:
`#[CommandAutoTag(name: 'foo:bar', desc: 'boo', hidden: true)]#`

The attribute could very well split the `hidden` and `aliases` settings apart - while the underlying code and pre-PHP8 apps would use the compact form, because dealing with many static properties + methods would be a maintenance pain imho.

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (jderusse) [inline src/Symfony/Component/Console/Command/LazyCommand.php:36]: when is it injected?

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/LazyCommand.php:36]: It's not. I added it for completeness, see https://github.com/symfony/symfony/pull/39851#issuecomment-761046995 for more.

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php
Comment (jderusse) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: Looking at Command::getDefaultName signature, $aliases can be null (we probably should add a test that cover this case)

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (jderusse) [inline src/Symfony/Component/Console/Command/LazyCommand.php:102]: Laziness is broken here because this is called by `Application::add` method here https://github.com/symfony/symfony/blob/29b88afeb717c3f993c4282a55679e5278084b9f/src/Symfony/Component/Console/Application.php#L493


ApplicationDescription::inspectApplication() -> Application::all() -> $this->has() -> $this->add()

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: null or not works the same for `explode()`, but I still added a fallback to the empty string

File: src/Symfony/Component/Console/Command/LazyCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/LazyCommand.php:102]: good catch, updated

**Review by jderusse — APPROVED** [review]

**Review by rheh — APPROVED** [review]

**Review by jderusse — APPROVED** [review]

**Review by wouterj — COMMENTED** [review]
Shouldn't this PR also update all internal commands to define the `$defaultDescription` property? Otherwise, you'll never have a lazy list command if I understand correctly.

**Comment by nicolas-grekas** [comment]
@wouterj good catch, PR updated. I also removed the "command" attributes in service definitions since they're not needed: the pass will read `$defaultName` instead.

File: src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php
Comment (jderusse) [inline src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php:64]: IMHO this could be done in the `Command::__construct` method:
- reduce bloating code
- prevent user's mistake that don't use the FrameworkBundle, and don't call `setDescription`

File: src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php
Comment (jderusse) [inline src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php:39]: logic of `isEnabled` should be duplicated in FrameworkExtension (to remove the tag) otherwise this command will always be enabled (same for all Command with `isEnabled` method)

File: src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php:64]: I don't want to bump the console requirement of all components, nor to change the signature of the constructor.

File: src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php:39]: good catch, `ServerLogCommand` is the only one, updated.

File: src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Monolog/Command/ServerLogCommand.php:64]: I still did what you suggested (calling setDescription()) in the constructor,
but I'm keeping all these calls to preserve compat with older versions of symfony/console

**Review by wouterj — APPROVED** [review]
I really like this move. While working on #38275 the non-lazy application describer was a real blocker, that's fixed now.

I do however agree with Javier that this convention based enabled setting is far from ideal. But there are enough months left to try out alternatives, so no need to block the merge imho.

File: src/Symfony/Component/Console/Command/Command.php
Comment (jderusse) [inline src/Symfony/Component/Console/Command/Command.php:79]: Didn't you removed this? 
If I understand well, it prevent a grand-child class to inherit the description from the child class. This could be an issue for the `defaultName` which should be unique, but I think this overhead is not needed in description.

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Command/Command.php:79]: Yes, and I added it back. This makes both properties behave the same, and not being inherited, which is the correct behavior to me.

**Comment by nicolas-grekas** [comment]
See https://github.com/symfony/maker-bundle/pull/792 for maker-bundle.

**Comment by nicolas-grekas** [comment]
Thank you all for the great review.

**Review by jderusse — APPROVED** [review]

**Pull request merged**
